core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,80 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(private val mapping: Map[Int, Int]) extends (Int => Int) {
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+
+  private lazy val inverseMapping = mapping.map(_.swap)
+
+  override def apply(k: Int): Int = mapping.getOrElse(k, k)
+
+  override def toString: String = {
+    mapping.toSeq.sorted
+      .map { case (k, v) => s"$k -> $v"}
+      .mkString("Perm(", ", ", ")")
+  }
+
+  def image: Set[Int] = mapping.keySet
+  def invert(k: Int): Int = inverseMapping.getOrElse(k, k)
+  def inverse: Perm = new Perm(inverseMapping)
+}
+
+object Perm {
+  def apply(mapping: Map[Int, Int]): Perm = {
+    new Perm(mapping.filter(Function.tupled(_ != _)))
+  }
+  def apply(pairs: (Int, Int)*): Perm = apply(Map(pairs: _*))
+
+  implicit val PermEq = new Eq[Perm] {
+    def eqv(p: Perm, r: Perm): Boolean = p.mapping == r.mapping
+  }
+  implicit val PermIntAction: Action[Int, Perm] = new PermIntAction
+  implicit val PermGroup: Group[Perm] = new PermGroup
+  implicit def PermSeqPartialAction[A, CC[A] <: SeqLike[A, CC[A]]](implicit cbf: Factory[A, CC[A]]): PartialAction[CC[A], Perm] = new PermSeqPartialAction[A, CC[A]]
+}
+
+final class PermIntAction extends Action[Int, Perm] {
+  def actr(k: Int, perm: Perm): Int = perm(k)
+  def actl(perm: Perm, k: Int): Int = perm.invert(k)
+}
+
+final class PermGroup extends Group[Perm] {
+  def empty: Perm = Perm(Map.empty[Int, Int])
+  def combine(x: Perm, y: Perm): Perm = Perm {
+    val preimages = x.image | y.image
+
+    preimages.foldLeft(Map.empty[Int, Int]) {
+      case (prevMap, preimage) =>
+        val image = y(x(preimage))
+        if (preimage != image) prevMap + ((preimage, image)) else prevMap
```

Comment:
Just noticed this conditional is now unnecessary.
This whole method could probably just become `Perm(preimages.map(pre => pre -> y(x(pre))).toMap)`